File: Questions 201 - 210/206_reverseLinkedList.cpp
#include<iostream>
#include<vector>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

ListNode* reverseList(ListNode* head) {
        
    // NULL or only head
    if(!head || !head->next)
        return head;
    
    // 2 nodes
    if(head->next->next == NULL) {
        head->next->next = head;
        head = head->next;
        head->next->next = NULL;
        return head;
    }
    
    // 3 nodes or more
    ListNode *p, *q, *r;
    p = head;
    q = head->next;
    r = head->next->next;

    // set next of first node to NULL
    p->next = NULL;
    
    while(r!= NULL){
        q->next = p;
        p = q;
        q = r;
        r = r->next;
    }
    q->next = p;
    head = q;
    
    return head;
}

int main() {
    ListNode *head = new ListNode(1);
    head->next = new ListNode(2);
    head->next->next = new ListNode(3);
    head->next->next->next = new ListNode(4);
    head->next->next->next->next = new ListNode(5);

    ListNode *p = head;
    cout<<"Original linked list: ";
    while(p != NULL) {
        cout<<p->val<<" ";
        p = p->next;
    }
    cout<<endl;

    p = reverseList(head);

    cout<<"Reversed linked list: ";
    while(p != NULL) {
        cout<<p->val<<" ";
        p = p->next;
    }
    cout<<endl;

    return 0;
}

File: Questions 201 - 300/Questions 201 - 210/206_reverseLinkedList.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def reverseList(head):

    # zero of one nodes
    if(head == None or head.next == None):
        return head

    # two nodes
    if(head.next.next == None):
        head.next.next = head
        head = head.next
        head.next.next = None
        return head

    # three nodes
    p = head
    q = head.next
    r = head.next.next

    # update the first node (which will now point to last)
    p.next = None

    while(r != None):
        q.next = p
        p = q
        q = r
        r = r.next

    # update last node
    q.next = p

    # update head
    head = q

    # return result
    return head


if __name__ == "__main__":

    head = ListNode(5)
    head.next = ListNode(10)
    head.next.next = ListNode(15)
    head.next.next.next = ListNode(20)
    head.next.next.next.next = ListNode(25)

    p = head
    print("Original list: ", end=" ")
    while(p != None):
        print(p.val, end=" ")
        p = p.next

    q = reverseList(head)

    print("\nReversed list: ", end=" ")
    while(q != None):
        print(q.val, end=" ")
        q = q.next
